pkg/manager/volumes/phase.go

```diff
@@ -0,0 +1,131 @@
+package volumes
+
+import (
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	klog "k8s.io/klog/v2"
+)
+
+type VolumePhase int
+
+const (
+	VolumePhaseUnknown VolumePhase = iota
+	// 1. isPVCRevisionChanged: false
+	// 2. needModify: true
+	// 3. waitForNextTime: true
+	VolumePhasePending
+	// 1. isPVCRevisionChanged: false
+	// 2. needModify: true
+	// 3. waitForNextTime: false
+	VolumePhasePreparing
+	// 1. isPVCRevisionChanged: true
+	// 2. needModify: true/false
+	// 3. waitForNextTime: false
+	VolumePhaseWaitForLeaderEviction
+	// 1. isPVCRevisionChanged: true
+	// 2. needModify: true/false
+	// 3. waitForNextTime: true
+	VolumePhaseModifying
+	// 1. isPVCRevisionChanged: false
+	// 2. needModify: false
+	// 3. waitForNextTime: true/false
+	VolumePhaseModified
+)
+
+func (p VolumePhase) String() string {
+	switch p {
+	case VolumePhasePending:
+		return "Pending"
+	case VolumePhasePreparing:
+		return "Preparing"
+	case VolumePhaseWaitForLeaderEviction:
+		return "WaitForLeaderEviction"
+	case VolumePhaseModifying:
+		return "Modifying"
+	case VolumePhaseModified:
+		return "Modified"
+	}
+
+	return "Unknown"
+}
+
+func (p *podVolModifier) getVolumePhase(vol *ActualVolume) VolumePhase {
+	if isPVCRevisionChanged(vol.PVC) {
+		if !p.waitForNextTime(vol.PVC, vol.Desired.StorageClass) {
+			return VolumePhaseWaitForLeaderEviction
+		}
+		return VolumePhaseModifying
+	}
+
+	if !needModify(vol.PVC, vol.Desired) {
+		return VolumePhaseModified
+	}
+
+	if p.waitForNextTime(vol.PVC, vol.Desired.StorageClass) {
+		return VolumePhasePending
+	}
+
+	return VolumePhasePreparing
+}
+
+func isPVCRevisionChanged(pvc *corev1.PersistentVolumeClaim) bool {
+	specRevision := pvc.Annotations[annoKeyPVCSpecRevision]
+	statusRevision := pvc.Annotations[annoKeyPVCStatusRevision]
+
+	return specRevision != statusRevision
+}
+
+func (p *podVolModifier) waitForNextTime(pvc *corev1.PersistentVolumeClaim, sc *storagev1.StorageClass) bool {
+	str, ok := pvc.Annotations[annoKeyPVCLastTransitionTimestamp]
+	if !ok {
+		return false
+	}
+	timestamp, err := time.Parse(time.RFC3339, str)
+	if err != nil {
+		return false
+	}
+	d := time.Since(timestamp)
+
+	m := p.getVolumeModifier(sc)
+
+	if m == nil {
+		return d < defaultModifyWaitingDuration
+	}
+
+	return d < m.MinWaitDuration()
+}
+
+func needModify(pvc *corev1.PersistentVolumeClaim, desired *DesiredVolume) bool {
+	size := desired.Size
+	scName := ""
+	if desired.StorageClass != nil {
+		scName = desired.StorageClass.Name
+	}
+
+	return isPVCStatusMatched(pvc, scName, size)
+}
+
+func isPVCStatusMatched(pvc *corev1.PersistentVolumeClaim, scName, size string) bool {
```

Comment:
any better name? the returned value is changed not matched. the same for `isPVCSpecMatched` in _pod_vol_modifier.go_